File: #1579.remove-max-number-of-edges-to-keep-graph-fully-traversable.cpp
class Solution {
public:
    int maxNumEdgesToRemove(int n, vector<vector<int>>& edges) {
        
        vector<int> a(n + 1);
        vector<int> b(n + 1);
        int compa;
        int compb;
        compa = n;
        compb = n;
        int edg = 0;
        for (int i = 0; i <= n; i++) {
            a[i] = b[i] = i;
        }

        for (auto i : edges) {
            if (i[0] == 3) {
                edg += Union(i[1], i[2], a, compa)
                | Union(i[1], i[2], b, compb);
            }
        }

        for (auto i : edges) {
            if (i[0] == 1) {
                edg += Union(i[1], i[2], a, compa);
            } else if(i[0] == 2) {
                edg += Union(i[1], i[2], b, compb);
            }
        }

        if (compa == 1 && compb == 1) {
            return edges.size() - edg;
        }
        return -1;
    }

private:
    int findP(int u,vector<int>& p) {

        if (p[u] == u) {
            return u;
        }

        return p[u] = findP(p[u],p);
    }

    bool Union(int a,int b,vector<int> &v,int &comp) {

        a = findP(a, v);
        b = findP(b, v);
        if (a == b) {
            return 0;
        }

        v[a] = b;
        comp--;
        return 1;
    }
};

File: #1579.remove-max-number-of-edges-to-keep-graph-fully-traversable.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number}
 */
var maxNumEdgesToRemove = function(n, edges) {
    let res = 0;
    let alice = Array(n+1).fill(-1);
    let bob = [...alice];
    for (let [type, x, y] of edges) {
        if (type === 3) {
            let m1 = union(x,y, alice);
            let m2 = union(x,y, bob)
            if (m1 && m2) {
                res++;
            }
        }
    }

    for (let [type, x, y] of edges) {
        if (type !== 3) {
            if (type == 1) {
                if (union(x,y,alice)) {
                    res++;
                }
            } else {
                if (union(x,y,bob)) {
                    res++;
                }
            }
        }
    }

    let minus1 = 0;
    for (let idx = 1; idx < n + 1; idx++) {
        if (alice[idx] === -1) {
            minus1++;
        }
    }

    if (minus1 > 1) {
        return -1;
    }

    for (let idx = 1; idx < n+1; idx++) {
        if (bob[idx] === -1) {
            minus1++;
        }
    }

    if (minus1 > 2) {
        return -1;
    }
    return res;

        function findPar(x, par) {
        if(par[x] == -1) {
            return x;
        }

        par[x] = findPar(par[x], par);
        return par[x];
    }

    function union(x,y,par) {
        let p_x = findPar(x,par);
        let p_y = findPar(y,par);
        if (p_x !== p_y) {
            par[p_x] = p_y;
            return false;
        }
        return true;
    }
};
